pcap/pcapio/ngread.go

```diff
@@ -0,0 +1,317 @@
+			intf.secondMask *= 10
+		}
+	}
+	intf.scaleDown = 1
+	intf.scaleUp = 1
+	if intf.secondMask < 1e9 {
+		intf.scaleUp = 1e9 / intf.secondMask
+	} else {
+		intf.scaleDown = intf.secondMask / 1e9
+	}
+	//XXX this list should be per section no?
+	r.ifaces = append(r.ifaces, intf)
+	return nil
+}
+
+// convertTime adds offset + shifts the given time value according to the given interface
+func (r *NgReader) convertTime(ifaceID int, ts uint64) (int64, int64) {
+	iface := r.ifaces[ifaceID]
+	return int64(ts/iface.secondMask + iface.TimestampOffset), int64(ts % iface.secondMask * iface.scaleUp / iface.scaleDown)
+}
+
+// readPacketHeader looks for a packet (enhanced, simple, or packet) and parses the header.
+// If an interface descriptor, an interface statistics block, or a section header is encountered, those are handled accordingly.
+// All other block types are skipped. New block types must be added here.
+func (r *NgReader) Read() ([]byte, BlockType, error) {
+	block := r.first
+	if block != nil {
+		r.first = nil
+		r.offset += uint64(len(block))
+		return block, TypeSection, nil
+	}
+	for {
+		typ, block, err := r.readBlock()
+		if err != nil {
+			return nil, 0, err
+		}
+		r.offset += uint64(len(block))
+		switch typ {
+		case ngBlockTypeEnhancedPacket:
+			if len(block) < 28 {
+				return nil, 0, ErrCorruptPcap
+			}
+			ifno := int(r.getUint32(block[8:12]))
+			if ifno >= len(r.ifaces) {
+				return nil, 0, fmt.Errorf("Interface id %d not present in section (have only %d interfaces)", r.ci.InterfaceIndex, len(r.ifaces))
+			}
+			caplen := int(r.getUint32(block[20:24]))
+			//length := int(r.getUint32(block[24:28]))
+			packet := block[28:]
+			if len(packet) < caplen {
+				return nil, 0, ErrCorruptPcap
+			}
+			return block, TypePacket, nil
+
+		case ngBlockTypeSimplePacket:
+			return nil, 0, errors.New("pcap-ng simple packets not supported")
+		case ngBlockTypeInterfaceDescriptor:
+			err := r.parseInterfaceDescriptor(block)
+			return block, TypeInterface, err
+		case ngBlockTypeInterfaceStatistics:
+			// ignore and drop
+		case ngBlockTypeSectionHeader:
+			err := r.parseSectionHeader(block)
+			return block, TypeSection, err
+		case ngBlockTypePacket:
+			return nil, 0, errors.New("pcap-ng deprecated type packet not supported")
+		}
+	}
+	return nil, 0, nil
+}
+
+// Interface returns interface information and statistics of interface with the given id.
+func (r *NgReader) Interface(i int) (NgInterface, error) {
+	if i >= len(r.ifaces) || i < 0 {
+		return NgInterface{}, fmt.Errorf("Interface %d invalid. There are only %d interfaces", i, len(r.ifaces))
+	}
+	return r.ifaces[i], nil
+}
+
+// NInterfaces returns the current number of interfaces.
+func (r *NgReader) NInterfaces() int {
+	return len(r.ifaces)
+}
+
+// Resolution returns the timestamp resolution of acquired timestamps before scaling to NanosecondTimestampResolution.
+func (r *NgReader) Resolution() gopacket.TimestampResolution {
+	//XXX
```

Comment:
If this isn't used, would you remove the method?

Reply:
ok